src/services/refactors/convertReactPureComponent.ts

```diff
@@ -0,0 +1,294 @@
+
+    const removeEmpty = (s: string) => s.replace(/\s/g, "");
+
+    type SFCLikeDeclaration =
+        | FunctionDeclaration
+        | FunctionExpression
+        | ArrowFunction;
+    function isSFCLikeDeclaration(node: Node): node is SFCLikeDeclaration {
+        return isFunctionDeclaration(node) || isFunctionExpression(node) || isArrowFunction(node);
+    }
+
+    interface Component {
+        name?: Identifier;
+        render: Block;
+        propsType?: TypeNode;
+        originNode: SFCLikeDeclaration | ClassLikeDeclaration;
+    }
+
+    /**
+     * This refactor follows this rule.
+     * * Is optional.
+     *
+     * 1. [x] Check if JSX Factory is React. If not, no actions will be provided
+     * 2. [x] If the selection is a function declaration of type `React.SFC`
+     *        [ ] Provide an action to convert it to `React.PureComponent`
+     *        [ ] * Make a function with 1 to 2 arguments (treat as props & context)
+     *              and explicit returnType T and T subtypeable to React.Element also convertible
+     * 3. [x] If the selection is a subclass of `React.Component` or `React.PureComponent`
+     *        [x] and if it satisifies `isConvertableClass()`
+     *        [ ] Provide an action to convert it to `React.SFC` or `React.PureComponent` (if it is `React.Component`)
+     * 4. [x] Resolve React by `getReferenceToReact()`
+     * 5. [ ] * Also support `propTypes`, `contextTypes`, `defaultProps`, `displayName` (Only Class -> Function)
+     * 6. [ ] * Also support convert of reference to `this.context` (in both ways)
+     */
+
+    registerRefactor(refactorName, {
+        getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+            // if (!checkJSXFactoryIsReact(context.host)) return;
+            if (!getReferenceToReact(context.file)) return;
+            const node = getSelectedNode(context);
+            if (!node) { return; }
+
+            if (isConvertibleReactClassComponent(node, context.file)) {
+                // const description = Diagnostics.???
+                const description = actionNameComponentToSFC;
+                return [{
+                    name: refactorName,
+                    description,
+                    actions: [{ description, name: actionNameComponentToSFC }]
+                }];
+            }
+            else if (isReactSFCDeclaration(node, context.file)) {
+                // const description = Diagnostics.???
+                const description = actionNameSFCToPureComponent;
+                return [{
+                    name: refactorName,
+                    description,
+                    actions: [{ description, name: actionNameSFCToPureComponent }]
+                }];
+            }
+        },
+        getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+            Debug.assert(actionName === actionNameComponentToSFC || actionName === actionNameSFCToPureComponent);

```

Comment:
you want to check again for the same conditions to defensively guard against changes to the file state before the actions are requested.

I would extract the code in `getAvailableActions` to `getConvertableReactElemnt` that will return you the SFC/Class or undefined if not found.

Reply:
See generateGetAccessorsAndSetAccessors.ts for a sample.